src/all.rs

```diff
@@ -634,33 +641,37 @@ unsafe impl<'a, Trait: ?Sized + TraitQuery> WorldQuery for All<&'a mut Trait> {
         state: &Self::State,
         access: &mut bevy_ecs::query::FilteredAccess<ComponentId>,
     ) {
+        let mut not_first = false;
+        let mut new_access = access.clone();
         for &component in &*state.components {
             assert!(
                 !access.access().has_write(component),
                 "&mut {} conflicts with a previous access in this query. Mutable component access must be unique.",
                 std::any::type_name::<Trait>(),
             );
-            access.add_write(component);
-        }
-    }
-
-    #[inline]
-    fn update_archetype_component_access(
-        state: &Self::State,
-        archetype: &bevy_ecs::archetype::Archetype,
-        access: &mut bevy_ecs::query::Access<bevy_ecs::archetype::ArchetypeComponentId>,
-    ) {
-        for &component in &*state.components {
-            if let Some(archetype_component_id) = archetype.get_archetype_component_id(component) {
-                access.add_write(archetype_component_id);
+            if not_first {
+                let mut intermediate = access.clone();
+                intermediate.add_write(component);
+                new_access.append_or(&intermediate);
+                new_access.extend_access(&intermediate);
+            } else {
+                new_access.add_write(component);
```

Comment:
It seems `add_write` now also sets the given component as "required", which is why lot of the entities in the tests have multiple components that implement the trait. The implementation of `WorldQuery` for `AnyOf` does `_new_access.required = _access.required.clone();` to solve this, but it is not possible here because `required` is private.

A way to fix this could be the following (not tested):

```suggestion
                new_access.and_with(component);
                new_access.access_mut().add_write(component);
```

This is essentially what `new_access.add_write(component)` does, except it doesn't set the bit in the `required` field.

Reply:
I don't think it is related to anything in the tests (they were all green at that point). I think the `required` field has to do with query transmutation. I don't quite understand how it works, but since that's what `AnyOf` does I assume your version is correct :+1: